Move `workOnTypes` to its own module

SquidDev opened the pull request:
Apologies for the PR spam here! This is the change I'm least certain about, so leaving it 'til last.

There's currently an import cycle between `Agda.TypeChecking.Reduce` and `Agda.TypeChecking.Irrelevance`. This is resolved by `.Reduce` `SOURCE`-importing `.Irrelevance`. However, this means that the definitions of `.Irrelevance` are not visible to `.Reduce`, meaning they cannot be inlined or specialised.

This becomes a bit of an issue with `workOnTypes`, which is called relatively frequently while reducing. As it is not specialised, it ends up allocating multiple closures for each call.

This moves `workOnTypes` (and several of the other context-modification functions) to a separate module, breaking the import cycle and allowing them to be inlined inside `.Reduce`. This yields a 1-3% speedup in typechecking.

---

Changing project layout for the purposes of optimisation may not be worth it, so obviously feel free to close this!

I'm also not entirely sure about the naming of this new module and which functions I've decided to move about. I'm afraid when we get onto irrelevance and modality is where my knowledge of type theory begins to break down a little, so definitely open to alternative suggestions here!

---

<details><summary>Performance measurements</summary>

Numbers were taken using GHC 9.2.7 and running Agda against `src/Data/Nat.lagda.md` from  https://github.com/plt-amy/1lab/commit/3b51679d95520c15ccd20accb9ab686e1eaa3c55. Agda was run under `perf stat` and with `+RTS -A128M -s -RTS --profile=internal`

### Master (791fb5e7b327d5e9b15146e5c234cfdbb6c5c14d)

```
Total                        32,597ms           
Miscellaneous                   122ms           
Typing                        1,317ms (22,605ms)
Typing.CheckRHS              10,147ms           
Typing.OccursCheck            5,761ms           
Typing.CheckLHS               1,504ms  (1,696ms)
Typing.CheckLHS.UnifyIndices    192ms           
Typing.TypeSig                1,594ms           
Typing.InstanceSearch         1,315ms           
Typing.Generalize               710ms           
Typing.With                      61ms           
Parsing                         385ms  (2,869ms)
Parsing.OperatorsExpr         1,983ms           
Parsing.OperatorsPattern        499ms           
Serialization                 1,530ms  (1,972ms)
Serialization.Compress          250ms           
Serialization.BinaryEncode      131ms           
Serialization.Sort               37ms           
Serialization.BuildInterface     22ms           
Scoping                         518ms  (1,530ms)
Scoping.InverseScopeLookup    1,011ms           
Deserialization                 671ms    (943ms)
Deserialization.Compaction      271ms           
Coverage                        925ms           
Highlighting                    548ms           
Positivity                      389ms           
DeadCode                        346ms           
Termination                     115ms    (217ms)
Termination.RecCheck             71ms           
Termination.Compare              16ms           
Termination.Graph                14ms           
Injectivity                      68ms           
Import                           58ms           
  60,075,994,392 bytes allocated in the heap
   2,025,767,248 bytes copied during GC
     102,845,184 bytes maximum residency (24 sample(s))
         509,128 bytes maximum slop
             343 MiB total memory in use (0 MB lost due to fragmentation)

                                     Tot time (elapsed)  Avg pause  Max pause
  Gen  0       429 colls,     0 par    2.940s   2.943s     0.0069s    0.0422s
  Gen  1        24 colls,     0 par    0.663s   0.663s     0.0276s    0.0835s

  TASKS: 4 (1 bound, 3 peak workers (3 total), using -N1)

  SPARKS: 0 (0 converted, 0 overflowed, 0 dud, 0 GC'd, 0 fizzled)

  INIT    time    0.004s  (  0.004s elapsed)
  MUT     time   29.011s  ( 28.905s elapsed)
  GC      time    3.602s  (  3.606s elapsed)
  EXIT    time    0.001s  (  0.006s elapsed)
  Total   time   32.619s  ( 32.521s elapsed)

  Alloc rate    2,070,797,507 bytes per MUT second

  Productivity  88.9% of total user, 88.9% of total elapsed


 Performance counter stats for 'scratch/agda-791fb5e7b327d5e9b15146e5c234cfdbb6c5c14d src/Data/Nat.lagda.md +RTS -A128M -s -RTS --profile=internal':

         32,612.34 msec task-clock:u                     #    1.002 CPUs utilized             
                 0      context-switches:u               #    0.000 /sec                      
                 0      cpu-migrations:u                 #    0.000 /sec                      
           108,090      page-faults:u                    #    3.314 K/sec                     
   132,493,868,462      cycles:u                         #    4.063 GHz                         (83.09%)
     8,691,989,623      stalled-cycles-frontend:u        #    6.56% frontend cycles idle        (83.39%)
    26,582,563,560      stalled-cycles-backend:u         #   20.06% backend cycles idle         (83.36%)
   134,893,320,440      instructions:u                   #    1.02  insn per cycle            
                                                  #    0.20  stalled cycles per insn     (83.38%)
    25,729,291,857      branches:u                       #  788.943 M/sec                       (83.40%)
     1,003,666,757      branch-misses:u                  #    3.90% of all branches             (83.39%)

      32.544316751 seconds time elapsed

      32.137447000 seconds user
       0.505074000 seconds sys
```

### This PR
```
Total                        31,614ms           
Miscellaneous                   121ms           
Typing                        1,322ms (21,623ms)
Typing.CheckRHS               9,775ms           
Typing.OccursCheck            5,397ms           
Typing.CheckLHS               1,484ms  (1,705ms)
Typing.CheckLHS.UnifyIndices    220ms           
Typing.TypeSig                1,494ms           
Typing.InstanceSearch         1,233ms           
Typing.Generalize               658ms           
Typing.With                      36ms           
Parsing                         404ms  (2,890ms)
Parsing.OperatorsExpr         1,998ms           
Parsing.OperatorsPattern        487ms           
Serialization                 1,647ms  (2,091ms)
Serialization.Compress          251ms           
Serialization.BinaryEncode      134ms           
Serialization.Sort               37ms           
Serialization.BuildInterface     19ms           
Scoping                         532ms  (1,546ms)
Scoping.InverseScopeLookup    1,013ms           
Deserialization                 661ms    (934ms)
Deserialization.Compaction      273ms           
Coverage                        851ms           
Highlighting                    504ms           
Positivity                      377ms           
DeadCode                        366ms           
Termination                      79ms    (185ms)
Termination.RecCheck             74ms           
Termination.Compare              16ms           
Termination.Graph                14ms           
Injectivity                      65ms           
Import                           54ms           
  57,661,616,040 bytes allocated in the heap
   2,106,164,560 bytes copied during GC
     121,974,352 bytes maximum residency (21 sample(s))
         483,760 bytes maximum slop
             386 MiB total memory in use (0 MB lost due to fragmentation)

                                     Tot time (elapsed)  Avg pause  Max pause
  Gen  0       414 colls,     0 par    2.915s   2.918s     0.0070s    0.0428s
  Gen  1        21 colls,     0 par    0.782s   0.782s     0.0372s    0.1137s

  TASKS: 4 (1 bound, 3 peak workers (3 total), using -N1)

  SPARKS: 0 (0 converted, 0 overflowed, 0 dud, 0 GC'd, 0 fizzled)

  INIT    time    0.002s  (  0.002s elapsed)
  MUT     time   27.933s  ( 27.825s elapsed)
  GC      time    3.697s  (  3.700s elapsed)
  EXIT    time    0.001s  (  0.004s elapsed)
  Total   time   31.634s  ( 31.530s elapsed)

  Alloc rate    2,064,259,767 bytes per MUT second

  Productivity  88.3% of total user, 88.2% of total elapsed


 Performance counter stats for 'scratch/agda-31d6429a609f49ac9de93e58893ece0e3b9036f3 src/Data/Nat.lagda.md +RTS -A128M -s -RTS --profile=internal':

         31,637.98 msec task-clock:u                     #    1.003 CPUs utilized             
                 0      context-switches:u               #    0.000 /sec                      
                 0      cpu-migrations:u                 #    0.000 /sec                      
           118,614      page-faults:u                    #    3.749 K/sec                     
   128,366,289,981      cycles:u                         #    4.057 GHz                         (83.09%)
     7,908,649,304      stalled-cycles-frontend:u        #    6.16% frontend cycles idle        (83.37%)
    26,302,224,166      stalled-cycles-backend:u         #   20.49% backend cycles idle         (83.38%)
   130,795,910,091      instructions:u                   #    1.02  insn per cycle            
                                                  #    0.20  stalled cycles per insn     (83.37%)
    24,704,004,435      branches:u                       #  780.834 M/sec                       (83.39%)
       994,350,025      branch-misses:u                  #    4.03% of all branches             (83.39%)

      31.556683401 seconds time elapsed

      31.165086000 seconds user
       0.495002000 seconds sys
```

</details>

andreasabel commented:
Thanks for the PR, @SquidDev !

I think the module `Agda.TypeChecking.Irrelevance` has outgrown its original scope (namely irrelevance), now it covers other modalities as well.
I looked at your separation of this module; it seems that not all functions that do not use reductions have been moved out.
I would suggest the split criterion to be: 
1. does not use reduction vs. (everything until including `wakeIrrelevantVars`?)
2. does use reduction (everything starting `UsableAtModality`?).

Could the module 1. maybe be named `Agda.TypeChecking.Monad.Modality` (rather than ... `Rules.WorkOnTypes`)?
The second could maybe keep the current name, unless there is some better suggestion.

SquidDev committed:
Move workOnTypes to its own module

This removes a dependency cycle between Reduce and workOnTypes, meaning
we no longer need to SOURCE import. This in turn allows workOnTypes to
be inlined/specialised, avoiding a common cause of allocations.

This provides a ~2% speedup in typechecking.

SquidDev commented:
Thank for for the feedback, that definitely feels a bit better!

andreasabel reviewed (approved):
Great, thanks!

This is definitely mergeable like this, but while you are at it, could you add a dedication (module haddock) for the new module `Agda.TypeChecking.Monad.Modality`?
Basically, explain what to expect to find in this module (what should go in there).

SquidDev committed:
Add a brief module description

SquidDev commented:
Oh, that readthedocs failure is very odd. Will try to reproduce locally, not sure what's caused that.

**Edit:** I've rebased on top of master, which will include the sphinx update. Hoping that will help at least!

andreasabel commented on src/full/Agda/TypeChecking/Monad/Modality.hs:
```diff
@@ -0,0 +1,200 @@
+the typing context. This module provides functions to update the context based
+on a given modality.
+
+See 'Agda.TypeChecking.Irrelevance'.
```
Module references need to be in double quotes in haddock, if I remember correctly.

SquidDev replied:
You're entirely right, thank you! It generates a link to the `Irrelevance` section of the (non-existent) `Agda.TypeChecking` module.

andreasabel reviewed (approved):
Thanks for the module comment!

Minor issue: There is a link to another module in single quotes, I think it need so be double quotes.  Could you check?
Unfortunately, I think CI does not have haddock preview---it builds the haddocks, but they do not get deployed somewhere.

SquidDev committed:
Fix module link
